### ISSUE TITLE
Rector for CakePHP repo

### ISSUE THREAD

**Issue by Harfusha** [cakephp/cakephp#17756]
### Description

Im working on updating this codebase using rector, i will supply PR for it when i will be done.

Should i add rector to composer dev section or should i keep only locally? 

### CakePHP Version

5.*, 4.*

**Comment by dereuromark** [comment]
The upgrade docs should explain that the tool should stay and be used as separate application 

See also
https://www.dereuromark.de/2023/09/28/cakephp-5-upgrade-guide/

**Comment by Harfusha** [comment]
No i'm running rector for THIS codebase. But i don't know if i should add something to composer.json or keep it locally + commit rector config i used

**Comment by dereuromark** [comment]
I guess keeping it locally for now is better.

**Comment by ADmad** [comment]
> No i'm running rector for THIS codebase.

To do what?

**Comment by Harfusha** [comment]
https://github.com/cakephp/cakephp/pull/17757

Something like this. If you approve it, i will fix the tests and styiling

**Comment by dereuromark** [comment]
You need to make smaller PRs with each only one specific change
Otherwise it is impossible to review or merge

**Comment by othercorey** [comment]
@Harfusha You have to explain the issues you are trying to fix not just that you are making changes.

Why did you start working on this?

**Comment by Harfusha** [comment]
I saw that there is a lot of things, rector can improve automatically. 

I made my PR just to show you what changes can be made and what is this issue about. (I will close my PR)

How should i split it? Per rector rule set or per code segment?

Or are this changes even worth the time? As i said, i made this PR for you to tell me if it is worth the time and if so an approach i should take.

**Comment by markstory** [comment]
We're unlikely to accept changes that modify whitespace/formatting. We already have established formatting standards and don't need another tool to manage that. We're also unlikely to accept changes that remove documentation annotation tags as we rely on those to build our API documentation.

Other types of changes like dead code removal, or improving usage of PHP stdlib where possible can be evaluated case by case.

**Comment by dereuromark** [comment]
> How should i split it? Per rector rule set or per code segment?

Try to split per rule set / type
As those can be easily discussed and approved as batch.

If you mix different things it will be possibly blocked by one of those.

**Comment by Harfusha** [comment]
Thanks for your instructions :) I will split it, and make only fixes, that are not affecting whitespace or formating :)

But what do you mean by type? Like Cache, Collection, Command, Console, etc? Or src, tests, etc?

I will try to make it today or tomorrow.

**Comment by dereuromark** [comment]
Rector type / rule

**Comment by ADmad** [comment]
Also avoid unnecessary renames. Like for e.g. changing `$e` to `$runtimeException` doesn't bring much value.

**Comment by Harfusha** [comment]
Can the PRs look like this?
https://github.com/cakephp/cakephp/pull/17758

**Comment by dereuromark** [comment]
Well, UP_TO_PHP_81 is technically not a rule, but a list of rules.
So still quite a big of different changes mixed together. All in all those do look OKish, with the exception of a few things:
- I wouldn't introduce readonly prop in the same changeset right now as the other changes
- `return array_map(fn ($key)=> (string)$key, $configurations);` is missing a CS rule, there should also be a space before the assignment arrow IMO
- test cases should not be touched this way adding \t - as this makes it only less readable.

Personally I find the fn ($key) thing way less readable, but that's maybe mainly also because I am not yet used to it at all.

**Comment by Harfusha** [comment]
Should i wait for the CS rule? I see in the PR that there is something about batch update later.

I will rollback readonly properties and the tabs :)

**Comment by Harfusha** [comment]
Next one i will do the CodeQuality rules.

All usable rules for us are touching 154 files. How should i split them so i dont have to make 59 PRs, and so i will not make 1 huge PR as the last one?

These are all usable rules for us.
- \Rector\CodeQuality\Rector\Assign\CombinedAssignRector::class
- \Rector\CodeQuality\Rector\BooleanAnd\RemoveUselessIsObjectCheckRector::class
- \Rector\CodeQuality\Rector\BooleanAnd\SimplifyEmptyArrayCheckRector::class
- \Rector\CodeQuality\Rector\BooleanNot\ReplaceMultipleBooleanNotRector::class
- \Rector\CodeQuality\Rector\BooleanNot\SimplifyDeMorganBinaryRector::class
- \Rector\CodeQuality\Rector\ClassConstFetch\ConvertStaticPrivateConstantToSelfRector::class
- \Rector\CodeQuality\Rector\ClassMethod\InlineArrayReturnAssignRector::class
- \Rector\CodeQuality\Rector\ClassMethod\LocallyCalledStaticMethodToNonStaticRector::class
- \Rector\CodeQuality\Rector\Class_\InlineConstructorDefaultToPropertyRector::class
- \Rector\CodeQuality\Rector\Class_\StaticToSelfStaticMethodCallOnFinalClassRector::class
- \Rector\CodeQuality\Rector\Equal\UseIdenticalOverEqualWithSameTypeRector::class
- \Rector\CodeQuality\Rector\Expression\InlineIfToExplicitIfRector::class
- \Rector\CodeQuality\Rector\Expression\TernaryFalseExpressionToIfRector::class
- \Rector\CodeQuality\Rector\For_\ForRepeatedCountToOwnVariableRector::class
- \Rector\CodeQuality\Rector\Foreach_\ForeachItemsAssignToEmptyArrayToAssignRector::class
- \Rector\CodeQuality\Rector\Foreach_\ForeachToInArrayRector::class
- \Rector\CodeQuality\Rector\Foreach_\SimplifyForeachToCoalescingRector::class
- \Rector\CodeQuality\Rector\Foreach_\UnusedForeachValueToArrayKeysRector::class
- \Rector\CodeQuality\Rector\FuncCall\ArrayMergeOfNonArraysToSimpleArrayRector::class
- \Rector\CodeQuality\Rector\FuncCall\CallUserFuncWithArrowFunctionToInlineRector::class
- \Rector\CodeQuality\Rector\FuncCall\ChangeArrayPushToArrayAssignRector::class
- \Rector\CodeQuality\Rector\FuncCall\CompactToVariablesRector::class
- \Rector\CodeQuality\Rector\FuncCall\InlineIsAInstanceOfRector::class
- \Rector\CodeQuality\Rector\FuncCall\IsAWithStringWithThirdArgumentRector::class
- \Rector\CodeQuality\Rector\FuncCall\RemoveSoleValueSprintfRector::class
- \Rector\CodeQuality\Rector\FuncCall\SetTypeToCastRector::class
- \Rector\CodeQuality\Rector\FuncCall\SimplifyFuncGetArgsCountRector::class
- \Rector\CodeQuality\Rector\FuncCall\SimplifyInArrayValuesRector::class
- \Rector\CodeQuality\Rector\FuncCall\SimplifyStrposLowerRector::class
- \Rector\CodeQuality\Rector\FuncCall\SingleInArrayToCompareRector::class
- \Rector\CodeQuality\Rector\FuncCall\UnwrapSprintfOneArgumentRector::class
- \Rector\CodeQuality\Rector\FunctionLike\SimplifyUselessVariableRector::class
- \Rector\CodeQuality\Rector\Identical\BooleanNotIdenticalToNotIdenticalRector::class
- \Rector\CodeQuality\Rector\Identical\FlipTypeControlToUseExclusiveTypeRector::class
- \Rector\CodeQuality\Rector\Identical\SimplifyArraySearchRector::class
- \Rector\CodeQuality\Rector\Identical\SimplifyBoolIdenticalTrueRector::class
- \Rector\CodeQuality\Rector\Identical\SimplifyConditionsRector::class
- \Rector\CodeQuality\Rector\Identical\StrlenZeroToIdenticalEmptyStringRector::class
- \Rector\CodeQuality\Rector\If_\CombineIfRector::class
- \Rector\CodeQuality\Rector\If_\CompleteMissingIfElseBracketRector::class
- \Rector\CodeQuality\Rector\If_\ConsecutiveNullCompareReturnsToNullCoalesceQueueRector::class
- \Rector\CodeQuality\Rector\If_\ShortenElseIfRector::class
- \Rector\CodeQuality\Rector\If_\SimplifyIfElseToTernaryRector::class
- \Rector\CodeQuality\Rector\If_\SimplifyIfNotNullReturnRector::class
- \Rector\CodeQuality\Rector\If_\SimplifyIfNullableReturnRector::class
- \Rector\CodeQuality\Rector\If_\SimplifyIfReturnBoolRector::class
- \Rector\CodeQuality\Rector\LogicalAnd\AndAssignsToSeparateLinesRector::class
- \Rector\CodeQuality\Rector\LogicalAnd\LogicalToBooleanRector::class
- \Rector\CodeQuality\Rector\New_\NewStaticToNewSelfRector::class
- \Rector\CodeQuality\Rector\NotEqual\CommonNotEqualRector::class
- \Rector\CodeQuality\Rector\NullsafeMethodCall\CleanupUnneededNullsafeOperatorRector::class
- \Rector\CodeQuality\Rector\Switch_\SingularSwitchToIfRector::class
- \Rector\CodeQuality\Rector\Switch_\SwitchTrueToIfRector::class
- \Rector\CodeQuality\Rector\Ternary\ArrayKeyExistsTernaryThenValueToCoalescingRector::class
- \Rector\CodeQuality\Rector\Ternary\NumberCompareToMaxFuncCallRector::class
- \Rector\CodeQuality\Rector\Ternary\SimplifyTautologyTernaryRector::class
- \Rector\CodeQuality\Rector\Ternary\SwitchNegatedTernaryRector::class
- \Rector\CodeQuality\Rector\Ternary\TernaryEmptyArrayArrayDimFetchToCoalesceRector::class
- \Rector\CodeQuality\Rector\Ternary\UnnecessaryTernaryExpressionRector::class

**Comment by rochamarcelo** [comment]
Who defined these code quality rules? Is this related in any way with php-fig?

I did a quick look at the list of the rules, what is the issuw with compact ?https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#compacttovariablesrector 

Also, Arrow function is good but sometimes it can be hard to read the code.

**Comment by rochamarcelo** [comment]
Are you guys planning to use Rector with CI? If yes which rules are going to be the standard for CakePHP?

**Comment by Harfusha** [comment]
Issue with the compact is that IDEs cannot tell you what variable it is, when using compact you have only the name of the variable, but if you want to find out what type it have or utilize IDE CTRL + click function to get to assignment, it is unusable.

Yeah i looked at all rules in Codequality category and remove the already denied changes

**Comment by rochamarcelo** [comment]
Proper phpdoc should help IDE understand the return of array, at least with PHPStorm autocomplete is available when using phpdoc.

**Comment by Harfusha** [comment]
but when you are looking at the compact and have the variable on the other end of function, you need  to find it using its name. If you would have it in array as suggested by rector, you can just hover over variable, This is not only affecting return types. You would use phpdoc for each compact? Even if you would you will still lose the ability to click trough the variable

**Comment by rochamarcelo** [comment]
Maybe PHPStorm has more features than other IDE but It does allows 'click-to-definition' on compact args

**Comment by Harfusha** [comment]
Made new PR for ruleset UP_TO_PHP_81: https://github.com/cakephp/cakephp/pull/17762

**Comment by Harfusha** [comment]
> Maybe PHPStorm has more features than other IDE but It does allows 'click-to-definition' on compact args

This is not allowed in VScode.
![image](https://github.com/user-attachments/assets/35e76b78-684d-4464-9e42-b8eca4f28204)
![image](https://github.com/user-attachments/assets/0d1be006-16bb-4100-abf5-e3b3c13a451e)

**Comment by Harfusha** [comment]
> Next one i will do the CodeQuality rules.
> 
> All usable rules for us are touching 154 files. How should i split them so i dont have to make 59 PRs, and so i will not make 1 huge PR as the last one?
> 
> These are all usable rules for us.
> 
> * \Rector\CodeQuality\Rector\Assign\CombinedAssignRector::class
> * \Rector\CodeQuality\Rector\BooleanAnd\RemoveUselessIsObjectCheckRector::class
> * \Rector\CodeQuality\Rector\BooleanAnd\SimplifyEmptyArrayCheckRector::class
> * \Rector\CodeQuality\Rector\BooleanNot\ReplaceMultipleBooleanNotRector::class
> * \Rector\CodeQuality\Rector\BooleanNot\SimplifyDeMorganBinaryRector::class
> * \Rector\CodeQuality\Rector\ClassConstFetch\ConvertStaticPrivateConstantToSelfRector::class
> * \Rector\CodeQuality\Rector\ClassMethod\InlineArrayReturnAssignRector::class
> * \Rector\CodeQuality\Rector\ClassMethod\LocallyCalledStaticMethodToNonStaticRector::class
> * \Rector\CodeQuality\Rector\Class_\InlineConstructorDefaultToPropertyRector::class
> * \Rector\CodeQuality\Rector\Class_\StaticToSelfStaticMethodCallOnFinalClassRector::class
> * \Rector\CodeQuality\Rector\Equal\UseIdenticalOverEqualWithSameTypeRector::class
> * \Rector\CodeQuality\Rector\Expression\InlineIfToExplicitIfRector::class
> * \Rector\CodeQuality\Rector\Expression\TernaryFalseExpressionToIfRector::class
> * \Rector\CodeQuality\Rector\For_\ForRepeatedCountToOwnVariableRector::class
> * \Rector\CodeQuality\Rector\Foreach_\ForeachItemsAssignToEmptyArrayToAssignRector::class
> * \Rector\CodeQuality\Rector\Foreach_\ForeachToInArrayRector::class
> * \Rector\CodeQuality\Rector\Foreach_\SimplifyForeachToCoalescingRector::class
> * \Rector\CodeQuality\Rector\Foreach_\UnusedForeachValueToArrayKeysRector::class
> * \Rector\CodeQuality\Rector\FuncCall\ArrayMergeOfNonArraysToSimpleArrayRector::class
> * \Rector\CodeQuality\Rector\FuncCall\CallUserFuncWithArrowFunctionToInlineRector::class
> * \Rector\CodeQuality\Rector\FuncCall\ChangeArrayPushToArrayAssignRector::class
> * \Rector\CodeQuality\Rector\FuncCall\CompactToVariablesRector::class
> * \Rector\CodeQuality\Rector\FuncCall\InlineIsAInstanceOfRector::class
> * \Rector\CodeQuality\Rector\FuncCall\IsAWithStringWithThirdArgumentRector::class
> * \Rector\CodeQuality\Rector\FuncCall\RemoveSoleValueSprintfRector::class
> * \Rector\CodeQuality\Rector\FuncCall\SetTypeToCastRector::class
> * \Rector\CodeQuality\Rector\FuncCall\SimplifyFuncGetArgsCountRector::class
> * \Rector\CodeQuality\Rector\FuncCall\SimplifyInArrayValuesRector::class
> * \Rector\CodeQuality\Rector\FuncCall\SimplifyStrposLowerRector::class
> * \Rector\CodeQuality\Rector\FuncCall\SingleInArrayToCompareRector::class
> * \Rector\CodeQuality\Rector\FuncCall\UnwrapSprintfOneArgumentRector::class
> * \Rector\CodeQuality\Rector\FunctionLike\SimplifyUselessVariableRector::class
> * \Rector\CodeQuality\Rector\Identical\BooleanNotIdenticalToNotIdenticalRector::class
> * \Rector\CodeQuality\Rector\Identical\FlipTypeControlToUseExclusiveTypeRector::class
> * \Rector\CodeQuality\Rector\Identical\SimplifyArraySearchRector::class
> * \Rector\CodeQuality\Rector\Identical\SimplifyBoolIdenticalTrueRector::class
> * \Rector\CodeQuality\Rector\Identical\SimplifyConditionsRector::class
> * \Rector\CodeQuality\Rector\Identical\StrlenZeroToIdenticalEmptyStringRector::class
> * \Rector\CodeQuality\Rector\If_\CombineIfRector::class
> * \Rector\CodeQuality\Rector\If_\CompleteMissingIfElseBracketRector::class
> * \Rector\CodeQuality\Rector\If_\ConsecutiveNullCompareReturnsToNullCoalesceQueueRector::class
> * \Rector\CodeQuality\Rector\If_\ShortenElseIfRector::class
> * \Rector\CodeQuality\Rector\If_\SimplifyIfElseToTernaryRector::class
> * \Rector\CodeQuality\Rector\If_\SimplifyIfNotNullReturnRector::class
> * \Rector\CodeQuality\Rector\If_\SimplifyIfNullableReturnRector::class
> * \Rector\CodeQuality\Rector\If_\SimplifyIfReturnBoolRector::class
> * \Rector\CodeQuality\Rector\LogicalAnd\AndAssignsToSeparateLinesRector::class
> * \Rector\CodeQuality\Rector\LogicalAnd\LogicalToBooleanRector::class
> * \Rector\CodeQuality\Rector\New_\NewStaticToNewSelfRector::class
> * \Rector\CodeQuality\Rector\NotEqual\CommonNotEqualRector::class
> * \Rector\CodeQuality\Rector\NullsafeMethodCall\CleanupUnneededNullsafeOperatorRector::class
> * \Rector\CodeQuality\Rector\Switch_\SingularSwitchToIfRector::class
> * \Rector\CodeQuality\Rector\Switch_\SwitchTrueToIfRector::class
> * \Rector\CodeQuality\Rector\Ternary\ArrayKeyExistsTernaryThenValueToCoalescingRector::class
> * \Rector\CodeQuality\Rector\Ternary\NumberCompareToMaxFuncCallRector::class
> * \Rector\CodeQuality\Rector\Ternary\SimplifyTautologyTernaryRector::class
> * \Rector\CodeQuality\Rector\Ternary\SwitchNegatedTernaryRector::class
> * \Rector\CodeQuality\Rector\Ternary\TernaryEmptyArrayArrayDimFetchToCoalesceRector::class
> * \Rector\CodeQuality\Rector\Ternary\UnnecessaryTernaryExpressionRector::class

I will make PR for each one. As i saw in last PR, i would need to revert lot of changes when they are multiple in single PR.

**Comment by Harfusha** [comment]
Now i will make changes from CodingStandard rules

**Comment by Harfusha** [comment]
Next i will make DeadCode rules

**Comment by Harfusha** [comment]
Next i will make EarlyReturn + InstanceOf rules

**Comment by Harfusha** [comment]
Lastly i did the TypeDeclaration rules + some that were missing

**Comment by Harfusha** [comment]
Thats a lot of PRs, it's almost impossible that i catched all coding standards + psalm issues and phpstan + phpunit errors.

I will go trough all opened PRs tomorrow and fix the issues i caused :)

But its all the rules. Should i attach the rector file in this issue if you want to implement it into the actions?

**Comment by ADmad** [comment]
Yes, please provide the rector config for all the changes you have submitted.

**Comment by Harfusha** [comment]
[rector.txt](https://github.com/user-attachments/files/16494156/rector.txt)
Cannot send it as php file, so i changed it to .txt

**Comment by Harfusha** [comment]
After all PRs will be closed or merged, i will send a final config, with all denied rules in ignore list

**Comment by Harfusha** [comment]
After this https://github.com/cakephp/cakephp/pull/17779 will be finished, i will run all approved rector rules again on 5.next branch so if rector find any fixes he can do because something was changed in some rule.

And i wanted to run rector for cake 4 too. Can i just make one PR with approved rules? (I will just change the PhpVersion from 8.1 to 7.4 and change ruleset UP_TO_PHP_84 to UP_TO_PHP_74)

**Comment by Harfusha** [comment]
> Who defined these code quality rules? Is this related in any way with php-fig?
> 
> I did a quick look at the list of the rules, what is the issuw with compact ?https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#compacttovariablesrector
> 
> Also, Arrow function is good but sometimes it can be hard to read the code.

I want to add some resources why compact is not very friendly feature.
https://3v4l.org/SsHBl
https://phpstan.org/r/dcef3299-748b-4719-821f-149ffd37e508 VS https://phpstan.org/r/5f876e2b-c66d-4397-add4-2b8598c581e6

**Comment by ADmad** [comment]
> And i wanted to run rector for cake 4 too.

4.x won't be getting a new minor release and I for one am not in favor of tinkering with a stable branch for very little gain.

**Comment by dereuromark** [comment]
> I want to add some resources why compact is not very friendly feature.

The performance topic can sound relevant. But if you look closely, those are very limited in numbers, usually 1-4 per controller action and thus request and as such not measurable.

The other issue is not relevant as we are doing this at the end of the controller action and from there on there is introspection afterwards.
The templates are not connected and as such your examples are void here.

**Comment by Harfusha** [comment]
> > And i wanted to run rector for cake 4 too.
> 
> 4.x won't be getting a new minor release and I for one am not in favor of tinkering with a stable branch for very little gain.

But there is milestone for it https://github.com/cakephp/cakephp/milestone/304

**Comment by dereuromark** [comment]
> > > And i wanted to run rector for cake 4 too.
> > 
> > 
> > 4.x won't be getting a new minor release and I for one am not in favor of tinkering with a stable branch for very little gain.
> 
> But there is milestone for it https://github.com/cakephp/cakephp/milestone/304

They were auto added here since those didn't fit a patch release.

**Comment by LordSimal** [comment]
so.... since all the associated PR's have either been merged or closed can this issue also be closed? Or is there something missing?

**Comment by Harfusha** [comment]
I will update the rector config and add it here when i get home and then i will close this issue

**Comment by Harfusha** [comment]
[rector.php.txt](https://github.com/user-attachments/files/16716782/rector.php.txt)
Here is the final rector config

**Comment by dereuromark** [comment]
It could probably be added to the static analyzer / CS part of CI?
As readonly check (not changing, but alerting with error code) if it detects anything?

**Comment by Harfusha** [comment]
Should i do it in this issue or create a new one?

**Comment by dereuromark** [comment]
I recommend opening up a new one referring to this one, as this here is already quite long. 
And CI for it is a separate follow up anyways.

**Comment by dereuromark** [comment]
We can also add a page to https://github.com/cakephp/cakephp/wiki explaining how to quickly run them if needed.
Since the CI part will only do the checking for now.
